src/ucp/tag/rndv.c

```diff
@@ -574,6 +686,90 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_bcopy_send, (self),
     return status;
 }
 
+UCS_PROFILE_FUNC(ucs_status_t, ucp_proto_progress_rndv_put_zcopy, (self),
+                 uct_pending_req_t *self)
+{
+    ucp_request_t *rndv_req = ucs_container_of(self, ucp_request_t, send.uct);
+    ucs_status_t status;
+    size_t offset, ucp_mtu, align, length;
+    ucp_rsc_index_t rsc_index;
+    const size_t max_iovcnt = 1;
+    uct_iov_t iov[max_iovcnt];
+    size_t iovcnt;
+    ucp_dt_state_t state;
+
+    if (ucp_ep_is_stub(rndv_req->send.ep)) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    if (!(ucp_tag_rndv_is_get_put_op_possible(rndv_req->send.ep, rndv_req->send.lane,
+                                              rndv_req->send.rndv_put.rkey_bundle.rkey))) {
+        /* can't perform put_zcopy - switch to AM rndv */
+        if (rndv_req->send.rndv_put.rkey_bundle.rkey != UCT_INVALID_RKEY) {
+            uct_rkey_release(&rndv_req->send.rndv_put.rkey_bundle);
+        }
+
+        /* switch to AM */
```

Comment:
Is it possible to choose the exact protocol by thresholds calculation before stating the request?